setuptools/dist.py

```diff
@@ -623,6 +623,21 @@ def _parse_config_files(self, filenames=None):  # noqa: C901
             except ValueError as e:
                 raise DistutilsOptionError(e) from e
 
+    def dash_to_underscore_warning(self, opt):
+        if opt in (
+            'home-page', 'download-url', 'author-email',
+            'maintainer-email', 'long-description', 'build-base',
+            'project-urls', 'license-file', 'license-files',
+            'long-description-content-type',
```

Comment:
How sure are we that this list is exhaustive? If we've missed one, what will happen? Did you consider just checking for `'-' in opt`? I know it was me who suggested the explicit list, so I'm critiquing my own guidance here too.

Reply:
I like the way you're thinking in 9c2b717.

Reply:
Hmm that is true. I wrote it with the assumption that we could continue adding to the list if needed, but on second thought that may not be very efficient or compatible. I wrote another fix which could be compatible with more users, which instead of listing the options to warn, it will generally warn if `'-' in opt` and for a certain list it won't do the conversion of hyphen to underscore so that it addresses the issues of PR #1608 for `extras_require` and `data_files`. So if more users report they need the dash, we can add to a smaller list. 

One concern I have with the new fix is that the warning can't distinguish between users who use hyphens meaningfully which may be a confusing error message for them. Although, I think it would also be more compatible for more users. What do you think?